[SUREFIRE-1254] add color messages

Tibor17 opened the pull request.

hboutemy commented on maven-failsafe-plugin/pom.xml:
```diff
@@ -127,6 +127,19 @@
             </configuration>
           </execution>
+          <execution>
+            <id>shared-logging-generated-sources</id>
```
IMHO, a little comment to explain why the source code is rebuilt here could be useful for future maintainers

Tibor17 replied:
Good point. I used m-shared-u:3.1.0 but there was compilation error I know from branch 3.0-rc1.
One class of m-shared-u appeared in another package related to report. I don't remember the name of the class but it can be easily reproduced.
This code would have very short life time because we already have branch `3.0-rc1` and one IT should be fixed and Surefire 2.x is gone.

hboutemy commented on maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/log/PluginConsoleLogger.java:
```diff
@@ -0,0 +1,126 @@
+    {
+        if ( mojoLogger.isDebugEnabled() )
+        {
+            mojoLogger.debug( message );
```
wow, the whole message with debug level color, and more generally each message with level color? this will give too much color.
I need to see the effective result, but if it's really what I fear, it was not the intent when adding color to Maven output: the intent is just to have a few colored words with general normal message (and nothing else than level with level color)

Tibor17 replied:
The purpose is to separate Mojo Logger.  I cannot use Mojo in forked JVM. Many times the code uses the same classes and interfaces in in-plugin tests or forked-jvm tests but the same classes are extended and behave different in in these two places however the super type is used eveywhere like polymorphism. The same happened with ConsoleLogger and it has advantage because all can be developed independently and the logger would be same (unlike sometimes we used System.out and Mojo logger or console logger), so now it's only one.
The important things is that `ConsoleLogger` does not need e.g. `isDebugEnabled` method but `AbstractSurefireMojo` used this for long time, so I decided to use concrete implementation class of `ConsoleLogger` in MOJO but that's the second reason to have this class here :)
Unfortunately we use Java 5; otherwise you would see those annotations in methods implemented from interface via `@Override`.

hboutemy replied:
here, I completely missed the point: ok, now I understand that it's just proxying log call from one API to the other

my fear is more on DefaultColoredConsoleMessageBuilder, where it seems to add color to each log message: I'll test

Tibor17 commented:
@hboutemy 
Hi Herve, I will amend the commit. So one commit would still appear however it is in progress. Surefire report has I think three place with logger only.
Maybe important to know. I use ConsoleLogger in with methods like `debug/info/warning/errror` in in-plugin jvm and forked jvm as well. This was here all the history of Surefire but I only extended with few more methods. The implementations should be different in forked-jvm and simply, as higher design principle was, the surefire providers are slaves and they only encode and send a message to main process or receive a command to do. The main process does what it has to do; either creates a report of displays message in console. From the same principle in-plugin jvm is the only place where the ClassLoader should see ANSI and MessageBuilder of m-shared-u. So this separation is safe because such classes will never be loaded or failed to load in forked jvm and this prevents me from having ugly jira issues. So I decided to use SPI to separate implementation which is in surefire-common from surefire-api which is used everywhere. The SPI is a builder of colored messages only.
So now we will have logs like `[INFO] Running MyTest` with colors instead of `Running MyTest`. 
Three advantages:
+ I fixed the issue you reported https://issues.apache.org/jira/browse/SUREFIRE-1254
+ many people reported bug that Surefire writes anything in consolve, so logger will again control the output and we don't have to add configuration parameter in plugin which is good, thus we fixed https://issues.apache.org/jira/browse/SUREFIRE-725 as well
+ and what I like is peace in code because as I told you before developers come and go to Surefire and everybody solved his problem without maintaining for long time. And that's the reason why logger was sometimes System.out, or System.err, or ConsoleLogger or mojo logger. I like when the code has internal principles that the attitude is just one; unlike use directly worked with stream PrintWriter which was wrapped stream of System.out which disallowed you to use normal logger like in Maven with slf4j and disallowed you to see messages like `[INFO] ...` etc.

Tibor17 commented:
@hboutemy I will remove the SPI. The code will be simple if I just move the implementation to `surefire-common`. I thought that `surefire-common` appeared in classpath of forked jvm but I see now the classpath and no such artifact is in there.

hboutemy commented on surefire-logger-api/src/main/java/org/apache/maven/plugin/surefire/log/api/Level.java:
```diff
@@ -0,0 +1,83 @@
+    public static Level resolveLevel(boolean hasSuccessful, boolean hasFailure, boolean hasError,
+                                     boolean hasSkipped, boolean hasFlake) {
+        boolean isRed = hasFailure | hasError;
+        if (isRed) {
```
you'll need to reformat this class (and NullConsoleLogger) since they don't use Maven formatting conventions then are detected by Checkstyle

Tibor17 replied:
yes, this has benn done - yet not committed.
I am currently running build and one before last module failed `surefire-setup-integration-tests` because of wrong shading.
The last module is `surefire-integration-tests` where I should switch off collors and then I have some activity to compare all binaries against binaries of previous release 2.19.1.
Then I will run manual testing and compare console with old version and several features which alter the console like `rerun`, `assumptions` and reports.
Maybe we should then test together.

hboutemy commented on surefire-api/src/main/java/org/apache/maven/surefire/booter/CommandReader.java:
```diff
@@ -46,10 +48,8 @@
-import static org.apache.maven.surefire.util.internal.StringUtils.isNotBlank;
-import static org.apache.maven.surefire.util.internal.StringUtils.isBlank;
 import static org.apache.maven.surefire.util.internal.DaemonThreadFactory.newDaemonThread;
+import static org.apache.maven.surefire.util.internal.StringUtils.*;
```
import static org.apache.maven.surefire.util.internal.StringUtils.encodeStringForForkCommunication;
import static org.apache.maven.surefire.util.internal.StringUtils.isBlank;
import static org.apache.maven.surefire.util.internal.StringUtils.isNotBlank;
import static org.apache.maven.surefire.util.internal.StringUtils.requireNonNull;

Tibor17 replied:
yes this is change in my local code as well.

hboutemy commented on surefire-booter/src/main/java/org/apache/maven/surefire/booter/ForkedBooter.java:
```diff
@@ -38,11 +38,16 @@
 import org.apache.maven.surefire.suite.RunResult;
 import org.apache.maven.surefire.testset.TestSetFailedException;
 
+import static java.lang.System.*;
```
I really don't like this static import nor the next one: System.out and System.err in code is a lot more manageable that out and err with static import

Tibor17 replied:
My IDEA added star. I should fix it in her configuration.

Tibor17 commented:
@hboutemy now we can test. All unit tests passed and I picked up first IT and passed successfully. So I am going to run all ITs. The content of binaries is expected and I compared it with previous release.

Tibor17 commented:
@hboutemy I found NPE so that consoleLogger is null. I will fix it and run the tests again. Some tests were successful.

hboutemy commented on maven-failsafe-plugin/src/main/java/org/apache/maven/plugin/failsafe/IntegrationTestMojo.java:
```diff
@@ -34,10 +34,11 @@
 import org.apache.maven.surefire.suite.RunResult;
 
+import static org.apache.maven.shared.utils.ReaderFactory.FILE_ENCODING;
+
```
is it really intentional to reformat everything to have static import on any static access? I don't find it convenient in general.

Tibor17 replied:
I use it for several purposes. When line is getting to lone around 120 characters and the constant is just a storage, or assertion statement when it is ugly to read the class; otherwise the whole statement is not verbal enough if class appears. I use IntelliJ IDEA which navigates me to the class where the constant is stored. I do not know how bout Eclipse abilities. I tried to use same name of method in static import and Java compiler failed, so it's safe compiler.
